analysis/quantifier-comparisons/12-spatial-transcriptomics.Rmd

```diff
@@ -0,0 +1,595 @@
+image_dir <- file.path(data_dir, "data", "images")
+
+# results directory 
+results_dir <- file.path(data_dir, "results")
+
+# create directories
+if(!dir.exists(data_dir)){
+  dir.create(data_dir, recursive = TRUE)
+}
+
+if(!dir.exists(image_dir)){
+  dir.create(image_dir, recursive = TRUE)
+}
+
+if(!dir.exists(results_dir)){
+  dir.create(results_dir, recursive = TRUE)
+}
+```
+
+
+```{r}
+mito_file <- file.path(base_dir, "sample-info", "Homo_sapiens.GRCh38.103.mitogenes.txt")
+  
+# read in mito genes 
+mito_genes <- readr::read_tsv(mito_file, col_names = "gene_id")
+mito_genes <- mito_genes %>%
+  dplyr::pull(gene_id) %>%
+  unique()
+```
+
+
+## Copy data from S3
+
+```{r}
+# grab alevin fry and cellranger output
+sample_ids <- c("SCPCR000372", "SCPCR000373")
+
+aws_copy_samples(local_dir = s3_out,
+                 s3_dir = "s3://nextflow-ccdl-results/scpca",
+                 samples = sample_ids,
+                 tools = c("alevin-fry-knee", "cellranger"))
+```
+
+## Comparing SpatialExperiment to Seurat
+
+### Importing Alevin-fry and Spaceranger output into a SpatialExperiment
+
+First we are testing use of a `SpatialExperiment` object to store the spatial transcriptomics results. 
+We will create two functions, one to deal with creating the combined object from both Alevin-fry and Spaceranger output, and the second to import Spaceranger output only. 
+
+```{r}
+# custom function for converting Alevin-fry and Spaceranger output to spe
+convert_fry_spaceranger_spe <- function(fry_dir, spaceranger_dir, sample_name){
+  fry_sce <- scpcaTools::read_alevin(fry_dir, 
+                                     usa_mode = TRUE,
+                                     which_counts = "spliced")
+  
+  # read in image data
+  image_data <- SpatialExperiment::readImgData(spaceranger_dir,
+                            sample_id = sample_name)
+  
+  # read in spatial coordinates 
+  spatial_data_file <- file.path(spaceranger_dir, "tissue_positions_list.csv")
+  spatial_data <- readr::read_csv(spatial_data_file, col_names = c("barcode", "in_tissue", "array_row",
+                                                                   "array_col","pxl_row_in_fullres",
+                                                                   "pxl_col_in_fullres"))
+  # find common cells between Alevin-fry and spaceranger 
+  spatial_data <- spatial_data %>%
+    dplyr::mutate(barcode = gsub("-1", "", barcode))
+  common_cells <- intersect(colnames(fry_sce), spatial_data$barcode)
+  
+  # subset fry_sce by common cells
+  fry_sce_subset <- fry_sce[,common_cells]
+  spatial_data_subset <- spatial_data %>%
+    dplyr::filter(barcode %in% common_cells)
+  
+  # construct 'SpatialExperiment'
+  fry_spe <- SpatialExperiment(
+    assays = list(counts = assay(fry_sce_subset)),
+    colData = colData(fry_sce_subset), 
+    imgData = image_data,
+    spatialData = DataFrame(spatial_data_subset),
+    spatialCoordsNames = c("pxl_col_in_fullres", "pxl_row_in_fullres"),
```

Comment:
Oh, I wonder if this is the place the coordinates get flipped. Usually rows before columns in R?

Reply:
This made me realize that this isn't the place where it's actually going wrong, it's in the plotting function, I'm plotting the columns as the x axis and the rows as the y axis, when it should probably be the other way around. This should also probably be flipped as well.

Reply:
I don't think it is the plotting function, or the spaceranger is plots would be the same, as they use the same function. I mean, I don't know which is correct, but what you have looks fine to me. Column coordinate as x makes sense for plotting, even if it is opposite the standard (row, column) order.